File: examples/javascript/ivf_index_example.mjs
/** IVF index -- inverted file with k-means clustering. Requires training data. */

import { randomUUID } from "node:crypto";
import { MembrainIvfIndex } from "membrain";

const DIMENSION = 128;
const NUM_TRAINING = 500;

function embed(seed) {
  const out = [];
  let state = seed * 1000 + 1;
  for (let i = 0; i < DIMENSION; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    out.push((state / 0x7fffffff) * 2 - 1);
  }
  return out;
}

const ids = Array.from({ length: NUM_TRAINING }, () => randomUUID());
const vectors = Array.from({ length: NUM_TRAINING }, (_, i) => embed(i));
const config = { dimension: DIMENSION, num_cells: 16, nprobe: 4, distance_metric: "Cosine" };

const index = MembrainIvfIndex.build(ids, vectors, config);

try {
  console.log(`Built IVF index: ${index.len()} vectors`);

  const newId = randomUUID();
  index.add(newId, embed(999));
  console.log(`After add: ${index.len()} vectors`);

  const results = index.search(embed(0.5), 5);
  for (const r of results) {
    console.log(`  id=${r.id.slice(0, 8)}... score=${r.score.toFixed(4)}`);
  }

  index.remove(newId);
  console.log(`After remove: ${index.len()} vectors`);
} finally {
  index.close();
}

File: examples/python/ivf_index_example.py
"""IVF index -- inverted file with k-means clustering. Requires training data."""

import random
import uuid

from membrain import MembrainIvfIndex

DIMENSION = 128
NUM_TRAINING = 500


def embed(seed: float) -> list[float]:
    random.seed(seed)
    return [random.gauss(0, 1) for _ in range(DIMENSION)]


def main() -> None:
    ids = [str(uuid.uuid4()) for _ in range(NUM_TRAINING)]
    vectors = [embed(seed=float(i)) for i in range(NUM_TRAINING)]

    config = {"num_cells": 16, "nprobe": 4, "distance_metric": "Cosine"}

    with MembrainIvfIndex.build(
        dimension=DIMENSION, ids=ids, vectors=vectors, config=config,
    ) as index:
        print(f"Built IVF index: {len(index)} vectors")

        new_id = str(uuid.uuid4())
        index.add(new_id, embed(seed=999.0))
        print(f"After add: {len(index)} vectors")

        results = index.search(embed(seed=0.5), k=5)
        for result in results:
            print(f"  id={result.id[:8]}... score={result.score:.4f}")

        index.remove(new_id)
        print(f"After remove: {len(index)} vectors")


if __name__ == "__main__":
    main()
